1-building-an-instagram-hashtag-typeahead-in-javascript/post.md

```diff
@@ -0,0 +1,202 @@
+    const activeWordOrHashtag = extractActiveWordOrHashtag(content, caretIndex);
+
+    // if the word that the user is editing is a hashtag, return it.
+    // otherwise, return null.
+    return activeWordOrHashtag[0] === '#' ? activeWordOrHashtag : null;
+};
+```
+You’ll notice that we have a supporting function, `extractActiveWordOrHashtag`, that is responsible for getting the word or hashtag that the user is editing from the post caption. Here’s how it looks:
+```js
+// Regex pattern that matches to a word or a hashtag.
+// Test it out here: [https://regex101.com/r/0Bl07o/2](https://regex101.com/r/0Bl07o/2)
+const hashtagOrWordRegex = /#*\w.*/g;
+
+// Gets the word that the user's caret is positioned on.
+const extractActiveWordOrHashtag = (content, caretIndex) => {
+    // First, backtrack until we find a character that can't
+    // be part of a word or hashtag.
+    let index = caretIndex;
+    let character = content[index];
+    do {
+      let matches = char.match(hashtagOrWordRegex);
+      // if this character is not part of a hashtag (e.g.
+      // it's a space or a period), return the word or
+      // hashtag in front of it.
+      if (!matches || !matches.length) {
+        return content
+           .slice(index + 1, content.length)
+           .match(hashtagOrWordRegex)[0];
+      }
+    // Otherwise, go to the previous character
+    index -= 1
+    } while (index > 0)
+}
+```
+## Rendering the Typeahead using React
+
+Thanks to our keydown handler and `getActiveHashtag`, the logic for whether or not to render the typeahead is in place. In our keydown handler, we return the result of `getActiveHashtag`. If this result isn’t `null`, we know that we need to render the typeahead; so we can pass `activeHashtag` as a prop to our `HashtagTypeahead` component and use it it in the render method like so:
+```js
+class HashtagTypeahead extends class {
+
+    ...
+
+    render () {
+        // If the user is editing a hashtag,
+        // render the typeahead to give the
+        // user suggestions for hashtags to use
+        if (this.props.activeHashtag) {
+            return (
+                <HashtagTypeaheadMenu
+                    activeHashtag={this.props.activeHashtag}
+                    hashtagOptions={...}
+                    onSelect={...)
+                />
+            )
+        }
+        // Otherwise, return nothing
+        return null;
+    }
+}
+```
+I abstracted away some of the React implementation details, because how you do this part depends on what typeahead library you want to use. At Tailwind, we use React extensively and have some component that do typeaheads for us already. If you’re looking for a good off-the-shelf typeahead components, check out [React Bootstrap Typeahead](https://www.npmjs.com/package/react-bootstrap-typeahead) or [React Autosuggest](https://react-autosuggest.js.org/). These might spare you the time and energy of building your own typeahead.
```

Comment:
would remove the tutorial tone here (unless i've missed that this tone is consistent throughout).